subsys/fs/fs.c

```diff
@@ -655,31 +655,62 @@ int fs_mount(struct fs_mount_t *mp)
 int fs_unmount(struct fs_mount_t *mp)
 {
 	int rc = -EINVAL;
+	size_t len = 0;
+	struct fs_mount_t *itr = NULL;
+	sys_dnode_t *node = NULL;
+
+	/* Unmount relies on mount path only */
+	if ((mp != NULL) && (mp->mnt_point != NULL)) {
+		len = strlen(mp->mnt_point);
+	}
 
-	if ((mp == NULL) || (mp->mnt_point == NULL) ||
-				(strlen(mp->mnt_point) <= 1)) {
+	if (len <= 1) {
 		LOG_ERR("invalid mount point!!");
 		return -EINVAL;
 	}
 
 	k_mutex_lock(&mutex, K_FOREVER);
-	if ((mp->fs == NULL) || mp->fs->unmount == NULL) {
-		LOG_ERR("fs ops functions not set!!");
-		rc = -EINVAL;
+
+	/* Check if mount point path is on the list of mounted file systems
+	 * first.
+	 */
+	SYS_DLIST_FOR_EACH_NODE(&fs_mnt_list, node) {
+		itr = CONTAINER_OF(node, struct fs_mount_t, node);
+		/* continue if length does not match */
+		if (len != itr->mountp_len) {
+			continue;
+		}
+
+		if (strncmp(mp->mnt_point, itr->mnt_point, len) == 0) {
+			break;
+		}
+	}
+
+	/* Mount path not found on list */
+	if (node == NULL) {
+		LOG_ERR("fs nothing mounted at %s", log_strdup(mp->mnt_point));
+		rc = -ENOENT;
```

Comment:
[Linux](https://man7.org/linux/man-pages/man2/umount.2.html) behavior is to return `-EINVAL` in this case.  Now that `-ENOTSUP` is distinguished my first thought is it's best to stick with that.

Reply:
>
> I'm not in principle opposed to storing mount points in a static array, but I'm not convinced it's a good thing, unless it's common to have more file systems available than you ever want simultaneously mounted.
> 
Unfortunately I do not understand what you mean here. I think that it is completely possible to have e.g. 3 mount points for 1 file system (e.g. FAT) to mount 4 different devices when you do not use all of them at the same time.
E.g you would have two SD cards sharing one mount point because you never use them at the same time so
you mount the one you need. The other two devices are mounted all the time.

> I have some hopes that in future zephyr most file systems will be described in devicetree, where their mount structure can be statically defined and need only be linked in or out of the active mount set. It'd be inconvenient to configure for 3 file systems and then break when a fourth one that's normally left unmounted is added at runtime (e.g. a pluggable device, or a partition dedicated to firmware updates, or temporary storage on an external high-capacity SRAM that's normally powered-off).

This is some good idea.

> 
> Also related to the fstab stuff: mount information should be not just the mount point (and the file system configuration referenced by pointer I think), but possibly other features like read-only state (dynamic) and automount flag (constant). IOW when you work up the devicetree bindings for the `fstab` stuff in #28320 I think `automount` should be in the fstab node, not the partition node.

I am working on these flags with the https://github.com/zephyrproject-rtos/zephyr/pull/27910 PR

Reply:
> 
> 
> [Linux](https://man7.org/linux/man-pages/man2/umount.2.html) behavior is to return `-EINVAL` in this case. Now that `-ENOTSUP` is distinguished my first thought is it's best to stick with that.

Change it to -`ENOTSUP`?

Reply:
No; I meant that a mis-use that formerly returned `-EINVAL` now returns `-ENOTSUP`, so the other errors can use `-EINVAL` without obscuring the case that's important to distinguish (that unmounting isn't supported).

I.e. if as-written this returns `-ENOENT` for a non-existent mount point (I think it does) it should instead return `-EINVAL`.

Reply:
> > I'm not in principle opposed to storing mount points in a static array, but I'm not convinced it's a good thing, unless it's common to have more file systems available than you ever want simultaneously mounted.
> 
> Unfortunately I do not understand what you mean here. I think that it is completely possible to have e.g. 3 mount points for ~1 file system (e.g. FAT) to~ mount 4 different devices when you do not use all of them at the same time.

Yes, that's what I meant (excluding the strike-out part; you'd need 4 files systems all of which could be FatFs).  I'm not convinced the RAM savings for mount point data that's only used sometimes is worth the potential faults if at runtime the number of simultaneous mounts exceeds the hard-coded limit.